scipio/src/executor.rs

```diff
@@ -322,28 +352,99 @@ impl LocalExecutor {
     /// ```
     pub fn new(binding: Option<usize>) -> io::Result<LocalExecutor> {
         let p = parking::Parker::new();
-        let le = LocalExecutor {
+        let mut le = LocalExecutor {
             queues: ExecutorQueues::new(),
             parker: p,
             binding,
+            id: EXECUTOR_ID.fetch_add(1, Ordering::Relaxed),
         };
 
-        if let Some(cpu) = binding {
-            bind_to_cpu(cpu)?;
-        }
+        le.init()?;
+        Ok(le)
+    }
 
-        let queues = le.queues.clone();
-        let index = 0;
+    /// Creates a single-threaded executor, optionally bound to a specific CPU, inside
+    /// a newly craeted thread. The parameter `name` specifies the name of the thread.
+    ///
+    /// This is a more ergonomic way to create a thread and then run an executor inside it
+    /// This function panics if creating the thread or the executor fails. If you need more
+    /// fine-grained error handling consider initializing those entities manually.
+    ///
+    /// You must call wait_on_executors() to wait for all executors spawned through spawn_new
+    /// to return
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scipio::LocalExecutor;
+    ///
+    /// // executor is a single thread, but not bound to any particular CPU.
+    /// LocalExecutor::spawn_new("myname", None, async move {
+    ///     println!("hello");
+    /// });
+    ///
+    /// LocalExecutor::wait_on_executors();
+    /// ```
+    pub fn spawn_new<F, T>(name: &'static str, binding: Option<usize>, fut: F)
```

Comment:
It is very surprising and counter intuitive that a `spawn_new` function would not return.

Reply:
yet it's interesting what a client would expect to be returned as far as it produces both: a thread, an executor and a fiber.

Reply:
The thing somewhat recalls:

```rust
        LocalExecutor::new(None)
            .and_then(|x: LocalExecutor| -> _ { Ok(x.run(async { println!("Hello") })) })
            .expect("good")
    )
```

Reply:
The only reasonable thing to return is a join handle.

Actually my preference would be to return an object that contains the join handle, and joins on it when it goes out of scope.

However I tested that approach. And it seems like the object can (and usually will) go out of scope before the end of the block, as soon as Rust deems it will not only be used. So that makes it essentially impossible to fire executors in parallel

Reply:
You are right that there is another way to write this Ivan, not like yours, a bit more complex (we are forcing a task queue, remember).

This is just a more ergonomic way